File: apps/api/app/schemas/feedback.py
import uuid
from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field
from app.models.feedback import FeedbackType


class FeedbackSubmitRequest(BaseModel):
    """Schema for submitting or updating recommendation feedback."""

    feedback_type: FeedbackType = Field(..., description="Structured feedback category")
    comment: Optional[str] = Field(None, max_length=1000, description="Optional text feedback or reason")


class FeedbackResponse(BaseModel):
    """Schema returning persisted recommendation feedback record."""

    id: uuid.UUID
    recommendation_event_id: uuid.UUID
    user_id: uuid.UUID
    feedback_type: FeedbackType
    comment: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class RecommendationEventResponse(BaseModel):
    """Schema returning recommendation event metadata."""

    id: uuid.UUID
    request_id: str
    user_id: uuid.UUID
    query: str
    entity_type: str
    entity_id: uuid.UUID
    rank_position: int
    relevance_score: float
    evidence_quality_score: float
    explanation_generated: Optional[str] = None
    graph_run_id: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


class EntityTypeMetrics(BaseModel):
    """Quality metrics grouped by entity type."""

    entity_type: str
    total_recommendations: int = 0
    feedback_count: int = 0
    helpful_count: int = 0
    negative_count: int = 0
    helpful_rate: float = 0.0
    average_relevance: float = 0.0
    average_evidence_quality: float = 0.0


class QualityMetricsResponse(BaseModel):
    """Aggregate recommendation quality and feedback metrics for admin reporting."""

    total_recommendations: int = 0
    total_feedback: int = 0
    helpful_count: int = 0
    negative_count: int = 0
    helpful_rate: float = 0.0
    negative_rate: float = 0.0
    explanation_coverage: float = 0.0
    evidence_quality_avg: float = 0.0
    feedback_coverage: float = 0.0
    average_relevance: float = 0.0
    by_entity_type: List[EntityTypeMetrics] = []


import enum

class QualityFlagEnum(str, enum.Enum):
    LOW_EVIDENCE = "LOW_EVIDENCE"
    NO_EVIDENCE = "NO_EVIDENCE"
    WEAK_SKILL_MATCH = "WEAK_SKILL_MATCH"
    WEAK_TECHNOLOGY_MATCH = "WEAK_TECHNOLOGY_MATCH"
    LOW_SEMANTIC_RELEVANCE = "LOW_SEMANTIC_RELEVANCE"
    NEGATIVE_USER_FEEDBACK = "NEGATIVE_USER_FEEDBACK"
    EXPLANATION_MISSING = "EXPLANATION_MISSING"
    OUTDATED_INFORMATION = "OUTDATED_INFORMATION"


class AdminRecommendationItem(BaseModel):
    """Item schema for administrative recommendation listing."""

    id: uuid.UUID
    request_id: str
    user_id: uuid.UUID
    query: str
    entity_type: str
    entity_id: uuid.UUID
    entity_title: Optional[str] = None
    rank_position: int
    relevance_score: float
    evidence_quality_score: float
    explanation_generated: Optional[str] = None
    feedback_type: Optional[FeedbackType] = None
    feedback_comment: Optional[str] = None
    quality_flags: List[str] = []
    created_at: datetime


class AdminRecommendationDetailResponse(BaseModel):
    """Safe detailed view of a recommendation event for administrative quality auditing."""

    id: uuid.UUID
    request_id: str
    user_id: uuid.UUID
    query: str
    entity_type: str
    entity_id: uuid.UUID
    entity_title: Optional[str] = None
    entity_subtitle: Optional[str] = None
    rank_position: int
    relevance_score: float
    evidence_quality_score: float
    explanation_generated: Optional[str] = None
    score_components: Dict[str, Any] = {}
    supporting_evidence: List[Dict[str, Any]] = []
    quality_flags: List[str] = []
    feedback: Optional[FeedbackResponse] = None
    created_at: datetime

File: apps/web/src/lib/api/feedback.ts
import { fetchApi } from "@/lib/api-client";

export type FeedbackType =
  | "HELPFUL"
  | "NOT_HELPFUL"
  | "PARTIALLY_HELPFUL"
  | "WRONG_MATCH"
  | "OUTDATED"
  | "INSUFFICIENT_EVIDENCE";

export interface FeedbackSubmitRequest {
  feedback_type: FeedbackType;
  comment?: string;
}

export interface FeedbackResponse {
  id: string;
  recommendation_event_id: string;
  user_id: string;
  feedback_type: FeedbackType;
  comment?: string;
  created_at: string;
  updated_at: string;
}

export interface EntityTypeMetrics {
  entity_type: string;
  total_recommendations: number;
  feedback_count: number;
  helpful_count: number;
  negative_count: number;
  helpful_rate: number;
  average_relevance: number;
  average_evidence_quality: number;
}

export interface QualityMetricsResponse {
  total_recommendations: number;
  total_feedback: number;
  helpful_count: number;
  negative_count: number;
  helpful_rate: number;
  negative_rate: number;
  explanation_coverage: number;
  evidence_quality_avg: number;
  feedback_coverage: number;
  average_relevance: number;
  by_entity_type: EntityTypeMetrics[];
}

export interface AdminRecommendationItem {
  id: string;
  request_id: string;
  user_id: string;
  query: string;
  entity_type: string;
  entity_id: string;
  entity_title?: string;
  rank_position: number;
  relevance_score: number;
  evidence_quality_score: number;
  explanation_generated?: string;
  feedback_type?: FeedbackType;
  feedback_comment?: string;
  quality_flags: string[];
  created_at: string;
}

export interface AdminRecommendationDetailResponse {
  id: string;
  request_id: string;
  user_id: string;
  query: string;
  entity_type: string;
  entity_id: string;
  entity_title?: string;
  entity_subtitle?: string;
  rank_position: number;
  relevance_score: number;
  evidence_quality_score: number;
  explanation_generated?: string;
  score_components: Record<string, unknown>;
  supporting_evidence: Array<Record<string, unknown>>;
  quality_flags: string[];
  feedback?: FeedbackResponse;
  created_at: string;
}

export const feedbackApi = {
  submitFeedback: async (recommendationId: string, data: FeedbackSubmitRequest): Promise<FeedbackResponse> => {
    return fetchApi<FeedbackResponse>(`/api/v1/feedback/recommendations/${recommendationId}`, {
      method: "POST",
      body: JSON.stringify(data),
    });
  },

  getFeedback: async (recommendationId: string): Promise<FeedbackResponse> => {
    return fetchApi<FeedbackResponse>(`/api/v1/feedback/recommendations/${recommendationId}`);
  },

  getMetrics: async (): Promise<QualityMetricsResponse> => {
    return fetchApi<QualityMetricsResponse>("/api/v1/admin/recommendations/metrics");
  },

  getAdminMetrics: async (): Promise<QualityMetricsResponse> => {
    return fetchApi<QualityMetricsResponse>("/api/v1/admin/recommendations/metrics");
  },

  getAdminRecommendations: async (options?: {
    query?: string;
    entity_type?: string;
    has_quality_flag?: string;
    page?: number;
    limit?: number;
  }): Promise<{ items: AdminRecommendationItem[]; total: number }> => {
    const params = new URLSearchParams();
    if (options?.query) params.append("query", options.query);
    if (options?.entity_type) params.append("entity_type", options.entity_type);
    if (options?.has_quality_flag) params.append("has_quality_flag", options.has_quality_flag);
    if (options?.page) params.append("page", options.page.toString());
    if (options?.limit) params.append("page_size", options.limit.toString());

    const queryString = params.toString();
    return fetchApi<{ items: AdminRecommendationItem[]; total: number }>(
      `/api/v1/admin/recommendations${queryString ? `?${queryString}` : ""}`
    );
  },

  getAdminRecommendationDetail: async (id: string): Promise<AdminRecommendationDetailResponse> => {
    return fetchApi<AdminRecommendationDetailResponse>(`/api/v1/admin/recommendations/${id}`);
  },
};
